### PR TITLE
LLM: add user guide for benchmarking

### PR DESCRIPTION
**Pull request by WeiguangHan**
Add GPU user guide for benchmarking.

### CODE REVIEW

**Comment by jason-dai** [comment]
Put it here https://github.com/intel-analytics/BigDL/tree/main/docs/readthedocs/source/doc/LLM/Quickstart, and rename it to sth like `benchmark_quickstart.md`

**Comment by WeiguangHan** [comment]
> Put it here https://github.com/intel-analytics/BigDL/tree/main/docs/readthedocs/source/doc/LLM/Quickstart, and rename it to sth like `benchmark_quickstart.md`

Sure. Will do it in next commit.

File: docs/readthedocs/source/_toc.yml [hunk 1]
```diff
@@ -52,6 +52,7 @@ subtrees:
                       - file: doc/LLM/Overview/KeyFeatures/inference_on_gpu
                       - file: doc/LLM/Overview/KeyFeatures/finetune
                       - file: doc/LLM/Overview/KeyFeatures/multi_gpus_selection
+                      - file: doc/LLM/Overview/Quickstart/benchmarking_quickstart
```
Comment (hkvision) [inline docs/readthedocs/source/_toc.yml:55]: benchmark_quickstart?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# Benchmarking
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:1]: BigDL-LLM Benchmarking

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+We can do benchmarking on Intel GPUs and CPUs. You can refer to [this](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install.html) to prepare the environment.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:3]: We can do benchmarking on Intel CPUs and GPUs using the benchmark script we provide.

Comment (hkvision) [reply docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:3]: prepare the environment for bigdl-llm.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+You can refer to [windows runtime configuration](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install_gpu.html#runtime-configuration) and [linux runtime configuration](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install_gpu.html#runtime-configuration) to source the needed environment variables for Intel GPU.
+
+
+## Run Script
+
+Run the script, wait for the model to finish running, and then you can obtain a CSV result file under the current folder.
+
+```
+cd your/local/workspace/BigDL/python/llm/dev/benchmark/all-in-one/
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:64]: put this in the clone repo part?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+python run.py
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:65]: Change to run-xxx.sh scripts for different platforms.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+Navigate to your local workspace and then download BigDL from GitHub.
+
+```
+cd your/local/workspace/
+git clone https://github.com/intel-analytics/BigDL.git
+```
+
+## Configure YAML File
+
+Replace the config.yaml under all-in-one folder with your own YAML file. The format of the content is as follows.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:16]: Modify the `config.yaml` under `all-in-one` folder for your own benchmark configurations. The contents of the yaml file are as follows:

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+- warm_up: Drop the results of the first time for better performance.
+- num_trials: Take the average of the three test result as the final performance result.
+- num_beams: Number of beams for beam search. 1 means no beam search.
+- low_bit: Convert the model to which precision you want to test.
+- batch_size: The number of samples on which the models makes predictions in one forward pass.
+- in_out_pairs: Inputs given to the model and expected outputs. 
+- test_api: Use different test functions on different machines.
+  - `transformer_int4_gpu` on Intel GPU for Linux
+  - `transformer_int4_gpu_win` on Intel GPU for Windows
+  - `deepspeed_transformer_int4_cpu` on Intel SPR Server
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:50]: remove this.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+cpu_embedding: False # whether put embedding to CPU (only avaiable now for gpu win related test_api)
+```
+
+The meanings of the different parameter configurations in the yaml file are shown below.
+
+
+- repo_id: The repo_id consists of two parts, one is the name of the model producer and the other is the name of the folder that actually holds the model data. 
+- local_model_hub: The folder path where the models are stored in your machine.
+- warm_up: Drop the results of the first time for better performance.
+- num_trials: Take the average of the three test result as the final performance result.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:42]: Can remove these two parameters? Users may not need to modify them?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+- test_api: Use different test functions on different machines.
+  - `transformer_int4_gpu` on Intel GPU for Linux
+  - `transformer_int4_gpu_win` on Intel GPU for Windows
+  - `deepspeed_transformer_int4_cpu` on Intel SPR Server
+  - `transformer_int4` on Intel CPU
+- cpu_embedding: Whether put embedding to the CPU (only avaiable now for windows gpu related test_api)
+
+## Runtime Configuration
+
+You can refer to [windows runtime configuration](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install_gpu.html#runtime-configuration) and [linux runtime configuration](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install_gpu.html#runtime-configuration) to source the needed environment variables for Intel GPU.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:56]: Remove this section.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+
+## Run Script
+
+Run the script, wait for the model to finish running, and then you can obtain a CSV result file under the current folder.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:61]: May add a detail description of the csv name?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md (hunk 6 above)
Comment (jason-dai) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:16]: which "all-in-one folder"?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+```
+
+The meanings of the different parameter configurations in the yaml file are shown below.
+
+
+- repo_id: The repo_id consists of two parts, one is the name of the model producer and the other is the name of the folder that actually holds the model data. 
+- local_model_hub: The folder path where the models are stored in your machine.
+- warm_up: Drop the results of the first time for better performance.
+- num_trials: Take the average of the three test result as the final performance result.
+- num_beams: Number of beams for beam search. 1 means no beam search.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:43]: remove num_beams

**Comment by hkvision** [comment]
Need to verify the CPU scripts in particular.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md (hunk 6 above)
Comment (hkvision) [reply docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:16]: After putting this cd in the clone repo section, it should be clear?
image

Comment (WeiguangHan) [reply docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:16]: `BigDL/python/llm/dev/benchmark/all-in-one`. Will describe it more clearly in next commit.

**Comment by Oscilloscope98** [comment]
Please add navigation to this guide here: https://github.com/intel-analytics/BigDL/blob/main/docs/readthedocs/source/_templates/sidebar_quicklinks.html#L17 and https://github.com/intel-analytics/BigDL/blob/main/docs/readthedocs/source/doc/LLM/Quickstart/index.rst

Please also update the TOC as we do here: https://github.com/intel-analytics/BigDL/blob/main/docs/readthedocs/source/_toc.yml#L42

File: docs/readthedocs/source/_toc.yml [hunk 12]
```diff
@@ -52,6 +52,7 @@ subtrees:
                       - file: doc/LLM/Overview/KeyFeatures/inference_on_gpu
                       - file: doc/LLM/Overview/KeyFeatures/finetune
                       - file: doc/LLM/Overview/KeyFeatures/multi_gpus_selection
+                      - file: doc/LLM/Overview/Quickstart/benchmark_quickstart
```
Comment (hkvision) [inline docs/readthedocs/source/_toc.yml:55]: Also need to modify https://bigdl.readthedocs.io/en/latest/doc/LLM/Quickstart/index.html

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+- cpu_embedding: Whether put embedding to the CPU (only avaiable now for windows gpu related test_api)
+
+
+## Run Script
+
+Run the script, wait for the model to finish running, and then you can obtain a CSV result file whose name includes `test_api` and the date of the day under the current folder. You can check if the test results in CSV meet your expectations, such as `1st token avg latency (ms)` and `2+ avg latency (ms/token)`.
+
+```
+# install necessary dependencies
+pip install tcmalloc
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:57]: tcmalloc installation is wrong, please refer to the README; and this is only for max-gpu

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+pip install pandas
+pip install omegaconf
+pip install wheel
+pip install einops
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:61]: are these two needed?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+## Run Script
+
+Run the script, wait for the model to finish running, and then you can obtain a CSV result file whose name includes `test_api` and the date of the day under the current folder. You can check if the test results in CSV meet your expectations, such as `1st token avg latency (ms)` and `2+ avg latency (ms/token)`.
+
+```
+# install necessary dependencies
+pip install tcmalloc
+pip install pandas
+pip install omegaconf
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:59]: put install dependencies in a separate section together with bigdl-llm installation.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+pip install wheel
+pip install einops
+
+# For ARC gpu, you can directly use `./run-arc.sh`.
+# For MAX gpu, you can directly use `./run-max-gpu.sh`.
+python run.py
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:65]: Use a selection bar for different platforms. Can add Arc, Max, iGPU for now, similar to here:
image

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+Navigate to your local workspace and then download BigDL from GitHub. Modify the `config.yaml` under `all-in-one folder` for your own benchmark configurations.
+
+```
+git clone https://github.com/intel-analytics/BigDL.git
+cd your/local/workspace/BigDL/python/llm/dev/benchmark/all-in-one/
+```
+
+## Configure YAML File
+
+The contents of the yaml file are as follows:
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:16]: can remove this sentence?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+in_out_pairs:
+  - '32-32'
+  - '1024-128'
+  - '2048-256'
+test_api:
+  - "transformer_int4_gpu"  # on Intel GPU
+cpu_embedding: False # whether put embedding to CPU (only avaiable now for gpu win related test_api)
+```
+
+The meanings of the different parameter configurations in the yaml file are shown below.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:36]: Some parameters in the yaml file that you can configure:

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+- repo_id: The repo_id consists of two parts, one is the name of the model producer and the other is the name of the folder that actually holds the model data. 
+- local_model_hub: The folder path where the models are stored in your machine.
+- low_bit: Convert the model to which precision you want to test.
+- batch_size: The number of samples on which the models makes predictions in one forward pass.
+- in_out_pairs: Inputs given to the model and expected outputs. 
+- test_api: Use different test functions on different machines.
+  - `transformer_int4_gpu` on Intel GPU for Linux
+  - `transformer_int4_gpu_win` on Intel GPU for Windows
+  - `transformer_int4` on Intel CPU
+- cpu_embedding: Whether put embedding to the CPU (only avaiable now for windows gpu related test_api)
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:48]: Whether __to__ put embedding __on__ CPU

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+num_beams: 1 # default to greedy search
+low_bit: 'sym_int4' # default to use 'sym_int4' (i.e. symmetric int4)
+batch_size: 1 # default to 1
+in_out_pairs:
+  - '32-32'
+  - '1024-128'
+  - '2048-256'
+test_api:
+  - "transformer_int4_gpu"  # on Intel GPU
+cpu_embedding: False # whether put embedding to CPU (only avaiable now for gpu win related test_api)
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:33]: remove the comments of these two lines?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+pip install tcmalloc
+pip install pandas
+pip install omegaconf
+pip install wheel
+pip install einops
+
+# For ARC gpu, you can directly use `./run-arc.sh`.
+# For MAX gpu, you can directly use `./run-max-gpu.sh`.
+python run.py
+```
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:66]: Need to add a description of the csv file.

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+The meanings of the different parameter configurations in the yaml file are shown below.
+
+
+- repo_id: The repo_id consists of two parts, one is the name of the model producer and the other is the name of the folder that actually holds the model data. 
+- local_model_hub: The folder path where the models are stored in your machine.
+- low_bit: Convert the model to which precision you want to test.
+- batch_size: The number of samples on which the models makes predictions in one forward pass.
+- in_out_pairs: Inputs given to the model and expected outputs.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:43]: this is not correct. Should be more detailed. Like input and output sequence length?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 23]
```diff
@@ -0,0 +1,66 @@
+# BigDL-LLM Benchmarking
+
+We can do benchmarking on Intel CPUs and GPUs using the benchmark script we provide. You can refer to [this](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install.html) to prepare the environment for bigdl-llm.
+
+## Clone Repo
+
+Navigate to your local workspace and then download BigDL from GitHub. Modify the `config.yaml` under `all-in-one folder` for your own benchmark configurations.
+
+```
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:9]: put cd your/local/workspace in a separate line at the beginning

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 24]
```diff
@@ -0,0 +1,107 @@
+# BigDL-LLM Benchmarking
+
+We can do benchmarking on Intel CPUs and GPUs using the benchmark script we provide.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:3]: do benchmarking for BigDL-LLM ...

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+
+         ./runpmax-gpu.sh
+
+      Please note that you need to run ``conda install -c conda-forge -y gperftools=2.10`` to install essential dependencies for Intel Data Center GPU Max.
+
+```
+
+## Result
+
+After the script runnning is completed, you can obtain a CSV result file under the current folder. You can mainly look at the results of columns `1st token avg latency (ms)` and `2+ avg latency (ms/token)`, check whether the column `actual input/output tokens` is consistent with the column `input/output tokens`, and the parameters you set before have been successfully applied in testing or not.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:107]: `2+ avg latency (ms/token)` for performance results. `Check ... ` in a separate sentence.
`and the parameters you set before have been successfully applied in testing or not.` this sentence seems not readable?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+         python run.py
+
+   .. tab:: Intel Arc™ A300-Series or Pro A60
+
+      .. code-block:: bash
+
+         set SYCL_CACHE_PERSISTENT=1
+         python run.py
+
+   .. tab:: Other Intel dGPU Series
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:79]: add (e.g. Arc™ A770)?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 27]
```diff
@@ -0,0 +1,113 @@
+# BigDL-LLM Benchmarking
+
+We can do benchmarking for BigDL-LLM on Intel CPUs and GPUs using the benchmark scripts we provide.
+
+## Prepare The Environment
+
+You can refer to [here](https://bigdl.readthedocs.io/en/latest/doc/LLM/Overview/install.html) to install the version of BigDL-LLM that you would like to benchmark based on your own machine. The following dependencies are also needed to run the benchmark scripts.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:7]: to install BigDL-LLM in your environment.

**Review by hkvision — APPROVED** [review]

File: docs/readthedocs/source/doc/LLM/Quickstart/index.rst [hunk 28]
```diff
@@ -8,4 +8,5 @@ BigDL-LLM Quickstart
 This section includes efficient guide to show you how to:
 
 * `Install BigDL-LLM on Windows with Intel GPU <./install_windows_gpu.html>`_
-* `Use Text Generation WebUI on Windows with Intel GPU <./webui_quickstart.html>`_
\ No newline at end of file
+* `Use Text Generation WebUI on Windows with Intel GPU <./webui_quickstart.html>`_
+* `Performance Benchmarking <./benchmark_quickstart.html>`_
```
Comment (Oscilloscope98) [inline docs/readthedocs/source/doc/LLM/Quickstart/index.rst:12]: We may need a verb phrase here? For example, "Conduct Performance Benchmarking with BigDL-LLM"?

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+
+## Configure YAML File
+
+```yaml
+repo_id:
+  - 'meta-llama/Llama-2-7b-chat-hf'
+local_model_hub: '/mnt/disk1/models'
+warm_up: 1
+num_trials: 3
+num_beams: 1 # default to greedy search
```
[the remark this answers is not in the document]
Comment (hkvision) [reply docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:32]: sure

File: docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+
+   .. tab:: Intel Data Center GPU Max
+
+      For Intel Data Center GPU Max Series, we recommend:
+
+      .. code-block:: bash
+
+         ./run-max-gpu.sh
+
+      Please note that you need to run ``conda install -c conda-forge -y gperftools=2.10`` to install essential dependencies for Intel Data Center GPU Max.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/LLM/Quickstart/benchmark_quickstart.md:107]: move this to before `./run-max-gpu.sh`

**Comment by hkvision** [comment]
Follow-up PR: https://github.com/intel-analytics/BigDL/pull/10373

**Pull request merged**